compiler_opt/es/blackbox_evaluator.py

```diff
@@ -64,30 +64,55 @@ class SamplingBlackboxEvaluator(BlackboxEvaluator):
   def __init__(self, train_corpus: corpus.Corpus,
                estimator_type: blackbox_optimizers.EstimatorType,
                total_num_perturbations: int, num_ir_repeats_within_worker: int):
-    self._samples = []
+    self._samples: list[list[corpus.LoadedModuleSpec]] = []
     self._train_corpus = train_corpus
     self._total_num_perturbations = total_num_perturbations
     self._num_ir_repeats_within_worker = num_ir_repeats_within_worker
     self._estimator_type = estimator_type
+    self._baselines: list[float | None] | None = None
 
     super().__init__(train_corpus)
 
-  def get_results(
-      self, pool: FixedWorkerPool,
-      perturbations: list[bytes]) -> list[concurrent.futures.Future]:
+  def _load_samples(self) -> None:
+    """Samples and loads modules if not already done.
+
+    Ensures self._samples contains the expected number of loaded samples.
+    Raises RuntimeError if loading fails and counts don't match.
+    """
     if not self._samples:
+      logging.info('Sampling and loading modules for evaluator...')
+
       for _ in range(self._total_num_perturbations):
-        sample = self._train_corpus.sample(self._num_ir_repeats_within_worker)
-        self._samples.append(sample)
+        samples = self._train_corpus.sample(self._num_ir_repeats_within_worker)
+        loaded_samples = [
+            self._train_corpus.load_module_spec(sample) for sample in samples
+        ]
+        self._samples.append(loaded_samples)
+
         # add copy of sample for antithetic perturbation pair
         if self._estimator_type == (
             blackbox_optimizers.EstimatorType.ANTITHETIC):
-          self._samples.append(sample)
+          self._samples.append(loaded_samples)
 
-    compile_args = zip(perturbations, self._samples)
+      logging.info('Done sampling and loading modules for evaluator.')
 
+      if self._estimator_type == (blackbox_optimizers.EstimatorType.ANTITHETIC):
+        expected_count = 2 * self._total_num_perturbations
+      else:
+        expected_count = self._total_num_perturbations
+
+      if len(self._samples) != expected_count:
+        raise RuntimeError("Some samples could not be loaded correctly.")
+
+  def _launch_compilation_workers(self,
+                                  pool: FixedWorkerPool,
+                                  perturbations: list[bytes] | None = None
+                                 ) -> list[concurrent.futures.Future]:
+    if perturbations is None:
+      perturbations = [None] * len(self._samples)
```

Comment:
What is the case where we launch workers with `None` perturbations?

Reply:
When we set the baseline in `set_baseline` method.